addons/wowl/static/src/core/download.ts

```diff
@@ -0,0 +1,140 @@
+import download from "../libs/download";
+import { _lt, Stringifiable } from "./localization";
+import parse from "../libs/content-disposition";
+import { Odoo } from "../types";
+
+interface DownloadFileOptionsFromForm {
+  type: "form";
+  form: HTMLFormElement;
+}
+
+interface DownloadFileOptionsFromParams {
+  type: "params";
+  url: string;
+  data: Map<string, string | Blob>;
+}
+
+interface DownloadFileOptionsCallbacks {
+  success?: () => void;
+  error?: (error: {
+    message: Stringifiable | String;
+    data: {
+      name?: Stringifiable | String;
+      title: Stringifiable | String;
+    };
+  }) => void;
+  complete?: () => void;
+}
+
+type DownloadFileOptions = (DownloadFileOptionsFromForm | DownloadFileOptionsFromParams) &
+  DownloadFileOptionsCallbacks;
+
+/**
+ * Cooperative file download implementation, for ajaxy APIs.
+ *
+ * Requires that the server side implements an httprequest correctly
+ * setting the `fileToken` cookie to the value provided as the `token`
+ * parameter. The cookie *must* be set on the `/` path and *must not* be
+ * `httpOnly`.
+ *
+ * It would probably also be a good idea for the response to use a
+ * `Content-Disposition: attachment` header, especially if the MIME is a
+ * "known" type (e.g. text/plain, or for some browsers application/json
+ *
+ * @param {Object} options
+ * @param {String} [options.url] used to dynamically create a form
+ * @param {Object} [options.data] data to add to the form submission. If can be used without a form, in which case a form is created from scratch. Otherwise, added to form data
+ * @param {HTMLFormElement} [options.form] the form to submit in order to fetch the file
+ * @param {Function} [options.success] callback in case of download success
+ * @param {Function} [options.error] callback in case of request error, provided with the error body
+ * @param {Function} [options.complete] called after both ``success`` and ``error`` callbacks have executed
+ * @returns {boolean} a false value means that a popup window was blocked. This
+ *   mean that we probably need to inform the user that something needs to be
+ *   changed to make it work.
+ */
+
+declare const odoo: Odoo;
+
+export function downloadFile(options: DownloadFileOptions) {
+  const xhr: XMLHttpRequest = new XMLHttpRequest();
+
+  let data: FormData;
+
+  if (options.type === "form") {
+    xhr.open(options.form.method, options.form.action);
+    data = new FormData(options.form);
+  } else {
+    xhr.open("POST", options.url!);
+    data = new FormData();
+    options.data.forEach((v, k) => {
+      data.append(k, v);
+    });
+  }
+
+  data.append("token", "dummy-because-api-expects-one");
+
+  if (odoo.csrf_token) {
```

Comment:
I use the odoo global var defined in views/templates.xml. It was done like that before. So we don't have to pass complex arguments to the function. Is it better, worse ?